server/src/test/java/org/opensearch/gateway/ReplicaShardBatchAllocatorTests.java

```diff
@@ -0,0 +1,849 @@
+     * the shard allocator to allocate it. There isn't a copy around to find anyhow.
+     */
+    public void testAsyncFetchWithNoShardOnIndexCreation() {
+        RoutingAllocation allocation = onePrimaryOnNode1And1Replica(
+            yesAllocationDeciders(),
+            Settings.EMPTY,
+            UnassignedInfo.Reason.INDEX_CREATED
+        );
+        testBatchAllocator.clean();
+        allocateAllUnassignedBatch(allocation);
+        assertThat(testBatchAllocator.getFetchDataCalledAndClean(), equalTo(false));
+        assertThat(testBatchAllocator.getShardEligibleFetchDataCountAndClean(), equalTo(0));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.UNASSIGNED).size(), equalTo(1));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.UNASSIGNED).get(0).shardId(), equalTo(shardId));
+    }
+
+    /**
+     * Verifies that for anything but index creation, fetch data ends up being called, since we need to go and try
+     * and find a better copy for the shard.
+     */
+    public void testAsyncFetchOnAnythingButIndexCreation() {
+        UnassignedInfo.Reason reason = RandomPicks.randomFrom(
+            random(),
+            EnumSet.complementOf(EnumSet.of(UnassignedInfo.Reason.INDEX_CREATED))
+        );
+        RoutingAllocation allocation = onePrimaryOnNode1And1Replica(yesAllocationDeciders(), Settings.EMPTY, reason);
+        testBatchAllocator.clean();
+        allocateAllUnassignedBatch(allocation);
+        assertThat("failed with reason " + reason, testBatchAllocator.getFetchDataCalledAndClean(), equalTo(true));
+        assertThat("failed with reason" + reason, testBatchAllocator.getShardEligibleFetchDataCountAndClean(), equalTo(1));
+    }
+
+    /**
+     * Verifies that when there is a full match (syncId and files) we allocate it to matching node.
+     */
+    public void testSimpleFullMatchAllocation() {
+        RoutingAllocation allocation = onePrimaryOnNode1And1Replica(yesAllocationDeciders());
+        DiscoveryNode nodeToMatch = randomBoolean() ? node2 : node3;
+        testBatchAllocator.addData(node1, "MATCH", null, new StoreFileMetadata("file1", 10, "MATCH_CHECKSUM", MIN_SUPPORTED_LUCENE_VERSION))
+            .addData(nodeToMatch, "MATCH", null, new StoreFileMetadata("file1", 10, "MATCH_CHECKSUM", MIN_SUPPORTED_LUCENE_VERSION));
+        allocateAllUnassignedBatch(allocation);
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).size(), equalTo(1));
+        assertThat(
+            allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(),
+            equalTo(nodeToMatch.getId())
+        );
+    }
+
+    /**
+     * Verifies that when there is a sync id match but no files match, we allocate it to matching node.
+     */
+    public void testSyncIdMatch() {
```

Comment:
how is it different from above test? Both can be checked with single test

Reply:
This test ensure that even if file checksum doesn't match but syncId is same, we still are allocating to matching node.
Kept both of these tests separate as we are doing in already existing ReplicaShardAllocatorTests
https://github.com/opensearch-project/OpenSearch/blob/91582e7c25a6f76eeebfa3ab74287178e1986491/server/src/test/java/org/opensearch/gateway/ReplicaShardAllocatorTests.java#L172-L202